pkg/policies/controlplane/components/differentiator.go

```diff
@@ -0,0 +1,147 @@
+
+// Type implements runtime.Component.
+func (*Differentiator) Type() runtime.ComponentType { return runtime.ComponentTypeSignalProcessor }
+
+// NewDifferentiator creates a differentiator component.
+func NewDifferentiator(diffProto *policylangv1.Differentiator) runtime.Component {
+	diff := &Differentiator{
+		window:      diffProto.Window.AsDuration(),
+		oldestIdx:   0,
+		newestIdx:   0,
+		capacity:    0,
+		initialized: false,
+	}
+	return diff
+}
+
+// NewDifferentiatorAndOptions creates a differentiator component and its fx options.
+func NewDifferentiatorAndOptions(diffProto *policylangv1.Differentiator, _ int, _ iface.Policy) (runtime.Component, fx.Option, error) {
+	return NewDifferentiator(diffProto), fx.Options(), nil
+}
+
+// Execute implements runtime.Component.Execute.
+func (d *Differentiator) Execute(inPortReadings runtime.PortToValue, tickInfo runtime.TickInfo) (runtime.PortToValue, error) {
+	if !d.initialized {
+		d.init(tickInfo)
+	}
+
+	inputVal := inPortReadings.ReadSingleValuePort("input")
+	outputVal := runtime.NewReading(0)
+
+	// add input to readings array
+	d.readings[d.newestIdx] = point{reading: inputVal, timestamp: tickInfo.Timestamp()}
+
+	if d.oldestIdx != d.newestIdx {
+		oldest := d.readings[d.oldestIdx]
+		newest := d.readings[d.newestIdx]
+
+		if !oldest.reading.Valid() {
+			oldest = d.firstValid(d.oldestIdx, true)
+		}
+
+		// find the newest valid reading for extrapolation
+		if !newest.reading.Valid() {
+			found := d.firstValid(d.newestIdx, false)
+			if found.reading.Valid() && oldest.reading.Valid() {
+				extrapolatedValue := d.extrapolate(oldest, found, tickInfo.Timestamp())
+				newest = point{reading: runtime.NewReading(extrapolatedValue), timestamp: tickInfo.Timestamp()}
+			}
+		}
+
+		// calculate the derivative
+		if oldest.reading.Valid() && newest.reading.Valid() {
+			diff := (newest.reading.Value() - oldest.reading.Value()) /
+				float64(newest.timestamp.UnixMilli()-oldest.timestamp.UnixMilli())
+			outputVal = runtime.NewReading(diff)
+		}
+	}
+
+	// shift pointers for newest and oldest
+	d.newestIdx = (d.newestIdx + 1) % d.capacity
+	if d.newestIdx == d.oldestIdx {
+		d.oldestIdx = (d.oldestIdx + 1) % d.capacity
+	}
+
+	return runtime.PortToValue{
+		"output": []runtime.Reading{outputVal},
+	}, nil
+}
+
+// DynamicConfigUpdate is a no-op for Differentiator.
+func (d *Differentiator) DynamicConfigUpdate(event notifiers.Event, unmarshaller config.Unmarshaller) {
+}
+
+func (d *Differentiator) extrapolate(firstVal, secondVal point, currentTime time.Time) float64 {
+	extValue := firstVal.reading.Value() + float64(currentTime.UnixMilli()-firstVal.timestamp.UnixMilli())/
+		float64(secondVal.timestamp.UnixMilli()-firstVal.timestamp.UnixMilli())*
+		(secondVal.reading.Value()-firstVal.reading.Value())
+	return extValue
+}
+
+func (d *Differentiator) init(tickInfo runtime.TickInfo) {
+	d.capacity = int(d.window / tickInfo.Interval())
```

Comment:
Instead of rounding down, take the ceiling similar to how we do it elsewhere, such as [EMA window](https://github.com/fluxninja/aperture/blob/main/pkg/policies/controlplane/components/ema.go#L62)

Reply:
ok